cli/src/cli_util.rs

```diff
@@ -3396,6 +3415,182 @@ impl CliRunner {
     }
 }
 
+/// TODO: refactor this according to opinion of maintainers
+///
+/// The problem is that the new dynamic completions from clap_complete want to
+/// be used via a completely separate code path from the normal program logic.
+/// There is good reason for this, see:
+/// https://docs.rs/clap_complete/latest/clap_complete/env/struct.CompleteEnv.html
+///
+/// However, it means that the dynamic completion functions don't have easy
+/// access to things like `ui` and `command_helper`. These are initialized
+/// directly in the `CliRunner::run` code path.
+///
+/// For now, I have simply duplicated the logic to get something working.
+///
+/// The long term solution should be discussed. The most obvious solution
+/// to me would be to extract the initialization logic that's shared between
+/// the normal run code path and the completion code path, and then pass the
+/// initialized objects as arguments to the run function. pseudo-code:
+///
+/// fn main() {
+///     let (ui, command_helper) = prepare_stuff();
+///     CliRunner::run(ui, command_helper);
+/// }
+///
+/// fn custom_completer() {
+///     let (ui, command_helper) = prepare_stuff();
+///     // do the thing
+/// }
```

Comment:
This is what I'm most unsure about. How to go about sharing this initialization logic between normal run path and completion logic?

It turns out I couldn't even copy-paste the initialization logic as-is. The completion logic cannot actually parse the CLI args, because the might still be invalid while the user/shell is requesting dynamic completions.

It seems tricky to share this logic in a nice way.

Reply:
I have no idea how `ArgValueCompleter` is supposed to access global resources. We might have to store them in static variables if completion is requested. Something like

1. process args normally until somewhere in `run_internal()`
2. if `$COMPLETE` is set, stop there and copy resources to static variables
3. `try_complete()`

It's ugly, but we'll need some static to handle extensions registered by caller?

Reply:
> I have no idea how `ArgValueCompleter` is supposed to access global resources.

From the perspective of `clap`, I don't think it's supposed to do that _at all_. The idea is for the completer functions to run on a completely separate code path from normal operation, which makes sense for most if not all applications.

Whenever I chose to use a static, I usually regretted it. What is the problem with extracting the shared logic into a function?

I'm also now thinking I might be approaching it completely wrong. I'm trying to access `ui` and `command_helper` while simply trying to get the list of local branches. Maybe I should use slightly lower-level APIs to implement completion logic? I'm not too familiar with the codebase yet, but theres a library part, maybe I should use that?

Reply:
> we'll need some static to handle extensions registered by caller?

What do you mean by that?

Reply:
> What extensions, registered how?

https://github.com/martinvonz/jj/blob/main/cli/examples/custom-backend/main.rs#L89

Maybe it can be registered at link time (with some static registration mechanisms), but it's currently constructed in call stack.

> Whenever I chose to use a static, I usually regretted it. 

I agree, but accessing `env::args()` is conceptually the same. They're bad, but work.

Reply:
> In what sense do such extensions interact with completions?

Yes. The `repo` would fail to load if the store backend weren't registered, for example. Google implements in-house backend for their monorepo.

Reply:
Aah, now I understand. The custom subcommands aren't the problem, but the storage backend is. I also can't see a different solution than a static variable in that case. I will try to implement that.

Reply:
Alright, done. Thanks for taking the time to review :bow: